pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -660,26 +660,39 @@ func (os *OpenStack) volumeService(forceVersion string) (volumeService, error) {
 		}
 		glog.V(3).Infof("Using Blockstorage API V2")
 		return &VolumesV2{sClient, os.bsOpts}, nil
+	case "v3":
+		sClient, err := os.NewBlockStorageV3()
+		if err != nil {
+			return nil, err
+		}
+		glog.V(3).Infof("Using Blockstorage API V3")
+		return &VolumesV3{sClient, os.bsOpts}, nil
 	case "auto":
-		// Currently kubernetes just support Cinder v1 and Cinder v2.
-		// Choose Cinder v2 firstly, if kubernetes can't initialize cinder v2 client, try to initialize cinder v1 client.
+		// Currently kubernetes support Cinder v1 / Cinder v2 / Cinder v3.
+		// Choose Cinder v3 firstly, if kubernetes can't initialize cinder v3 client, try to initialize cinder v2 client.
+		// If kubernetes can't initialize cinder v2 client, try to initialize cinder v1 client.
 		// Return appropriate message when kubernetes can't initialize them.
-		// TODO(FengyunPan): revisit 'auto' after supporting Cinder v3.
-		sClient, err := os.NewBlockStorageV2()
+		sClient, err := os.NewBlockStorageV3()
```

Comment:
Instead of making this more deeply nested over time, unfold it so it still reads linearly:

```golang
if sClient, err := os.NewBlockStorageV3(); err == nil {
  glog.V(3).Infof("Using Blockstorage API V3")
  return &VolumesV3{sClient, os.bsOpts}, nil
}

if sClient, err := os.NewBlockStorageV2(); err == nil {
  glog.V(3).Infof("Using Blockstorage API V2")
  return &VolumesV2{sClient, os.bsOpts}, nil
}

// ditto for V1

err_txt := "BlockStorage API version autodetection failed.  ..."
return nil, errors.New(err_txt)
```

Reply:
Nice, updated it.